File: src/skills/analyzeSalesOpportunity.js
/**
 * Analyse a flight search result and generate upsell / sales recommendations.
 *
 * This is a PURE LOGIC skill — no I/O.
 *
 * @param {object} params
 * @param {object|null} params.bestFlight        Best flight from SerpApi
 * @param {object|null} params.cheapestFlight    Cheapest flight from SerpApi
 * @param {string}      params.destination       IATA destination code
 * @param {Array}       [params.historicalPrices] Past prices from getRecentSearches
 * @returns {{ opportunityLevel: string, suggestions: string[], priceAnalysis: object }}
 */
export default function analyzeSalesOpportunity({
  bestFlight,
  cheapestFlight,
  destination,
  historicalPrices = [],
}) {
  const priceLow = parseInt(process.env.PRICE_THRESHOLD_LOW ?? '300', 10);
  const priceMedium = parseInt(process.env.PRICE_THRESHOLD_MEDIUM ?? '600', 10);

  const price = cheapestFlight?.price ?? bestFlight?.price ?? null;

  if (price === null) {
    return {
      opportunityLevel: 'unknown',
      suggestions: ['Não foi possível determinar o preço. Verificar manualmente.'],
      priceAnalysis: { price: null, thresholds: { low: priceLow, medium: priceMedium } },
    };
  }

  const suggestions = [];
  let opportunityLevel;

  if (price <= priceLow) {
    // ── Great deal — high opportunity ──
    opportunityLevel = 'high';
    suggestions.push(
      `✈️ Voo muito barato (${formatCurrency(price)})! Oportunidade de fechar rápido.`,
      `🏨 Sugerir upgrade de quarto — o cliente está economizando no aéreo.`,
      `🍾 Oferecer pacote de experiência premium (spa, jantar, transfer privativo).`,
      `📦 Propor combo voo + hotel com margem extra para o resort.`,
    );
  } else if (price <= priceMedium) {
    // ── Fair price — medium opportunity ──
    opportunityLevel = 'medium';
    suggestions.push(
      `✈️ Preço justo (${formatCurrency(price)}). Boa janela de venda.`,
      `🏨 Oferecer upgrade de categoria com desconto de cortesia.`,
      `🚗 Sugerir transfer aeroporto–hotel como valor agregado.`,
    );
  } else {
    // ── Expensive — lower opportunity, but still worth offering value ──
    opportunityLevel = 'low';
    suggestions.push(
      `✈️ Voo com preço elevado (${formatCurrency(price)}).`,
      `💡 Destacar flexibilidade de datas para buscar tarifas menores.`,
      `🎁 Oferecer crédito no resort como incentivo de fechamento.`,
    );
  }

  // ── Historical price comparison ──
  const historicalAnalysis = analyseHistoricalPrices(price, historicalPrices);
  if (historicalAnalysis.suggestion) {
    suggestions.push(historicalAnalysis.suggestion);
  }

  // Upgrade opportunity level if history shows this is a great deal
  if (historicalAnalysis.percentDiff != null && historicalAnalysis.percentDiff <= -15) {
    if (opportunityLevel === 'medium') opportunityLevel = 'high';
    if (opportunityLevel === 'low') opportunityLevel = 'medium';
  }

  // ── Destination-specific tips ──
  const destinationTips = getDestinationTip(destination);
  if (destinationTips) {
    suggestions.push(destinationTips);
  }

  return {
    opportunityLevel,
    suggestions,
    priceAnalysis: {
      price,
      currency: 'BRL',
      thresholds: { low: priceLow, medium: priceMedium },
      bestFlightPrice: bestFlight?.price ?? null,
      cheapestFlightPrice: cheapestFlight?.price ?? null,
      savingsVsBest:
        bestFlight?.price && cheapestFlight?.price
          ? bestFlight.price - cheapestFlight.price
          : null,
      historical: historicalAnalysis,
    },
  };
}

// ── Helpers ─────────────────────────────────────────────────

/**
 * Compare current price against historical average.
 */
function analyseHistoricalPrices(currentPrice, historicalPrices) {
  const validPrices = historicalPrices
    .map((h) => h.price)
    .filter((p) => p != null && !Number.isNaN(p));

  if (validPrices.length === 0) {
    return { averagePrice: null, percentDiff: null, suggestion: null };
  }

  const averagePrice = Math.round(
    validPrices.reduce((sum, p) => sum + p, 0) / validPrices.length,
  );

  const percentDiff = Math.round(
    ((currentPrice - averagePrice) / averagePrice) * 100,
  );

  let suggestion = null;

  if (percentDiff <= -15) {
    suggestion = `📉 Preço ${Math.abs(percentDiff)}% abaixo da média histórica (${formatCurrency(averagePrice)}) — ótima oportunidade!`;
  } else if (percentDiff <= -5) {
    suggestion = `📊 Preço ${Math.abs(percentDiff)}% abaixo da média (${formatCurrency(averagePrice)}) — momento favorável.`;
  } else if (percentDiff >= 15) {
    suggestion = `📈 Preço ${percentDiff}% acima da média (${formatCurrency(averagePrice)}) — sugerir datas alternativas.`;
  } else if (percentDiff >= 5) {
    suggestion = `📊 Preço ${percentDiff}% acima da média (${formatCurrency(averagePrice)}) — dentro da faixa normal.`;
  }

  return { averagePrice, percentDiff, suggestion, sampleSize: validPrices.length };
}

function formatCurrency(value) {
  return new Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL',
  }).format(value);
}

/**
 * Return a contextual tip based on known popular destinations.
 */
function getDestinationTip(iata) {
  const tips = {
    MIA: '🌴 Miami — sugerir passeio de barco em Biscayne Bay.',
    MCO: '🎢 Orlando — oferecer ingressos de parques como add-on.',
    CDG: '🗼 Paris — sugerir experiência gastronômica ou cruzeiro no Sena.',
    FCO: '🏛️ Roma — oferecer tour guiado pelo Vaticano.',
    LIS: '🇵🇹 Lisboa — sugerir passeio de tuk-tuk + degustação de vinhos.',
    CUN: '🏖️ Cancún — sugerir pacote all-inclusive com upgrade.',
    DXB: '🕌 Dubai — oferecer desert safari ou passeio de helicóptero.',
    JFK: '🗽 Nova York — oferecer ingressos para Broadway.',
    LHR: '🇬🇧 Londres — sugerir afternoon tea experience.',
  };
  return tips[iata?.toUpperCase()] ?? null;
}

File: app/skills/analyze_sales_opportunity.py
"""
skills/analyze_sales_opportunity.py — Pure logic: no I/O.

Classifies flight price into opportunity level (high/medium/low)
and generates upsell suggestions for the sales team.
"""

import os


def analyze_sales_opportunity(
    best_flight: dict | None,
    cheapest_flight: dict | None,
    destination: str,
    historical_prices: list | None = None,
) -> dict:
    """
    Returns { opportunity_level, suggestions, price_analysis }.

    opportunity_level: "high" | "medium" | "low" | "unknown"
    """
    price_low = float(os.environ.get("PRICE_THRESHOLD_LOW", 300))
    price_medium = float(os.environ.get("PRICE_THRESHOLD_MEDIUM", 600))

    price = (
        (cheapest_flight or {}).get("price")
        or (best_flight or {}).get("price")
    )

    if price is None:
        return {
            "opportunity_level": "unknown",
            "suggestions": ["Não foi possível determinar o preço. Verificar manualmente."],
            "price_analysis": {"price": None, "thresholds": {"low": price_low, "medium": price_medium}},
        }

    price = float(price)
    suggestions = []

    if price <= price_low:
        opportunity_level = "high"
        suggestions += [
            f"🎯 *Insight:* Tarifa extremamente competitiva detected ({_fmt(price)}).",
            "💰 *Estratégia:* O cliente possui alta margem de economia. Priorizar oferta de upgrade para suíte premium ou reserva de 1 noite extra.",
            "✨ *Incentivo:* Propor pacote 'Experience' (jantar romântico ou SPA) como cortesia pelo fechamento imediato.",
        ]
    elif price <= price_medium:
        opportunity_level = "medium"
        suggestions += [
            f"📊 *Insight:* Preço dentro da expectativa de mercado ({_fmt(price)}).",
            "🔄 *Estratégia:* Gatilho de escassez — informar que restam poucas vagas no hotel para o período, ancorando no valor do voo.",
            "🚗 *Valor Agregado:* Cortesia de transfer round-trip aeroporto-hotel pode ser o diferencial para conversão agora.",
        ]
    else:
        opportunity_level = "low"
        suggestions += [
            f"⚠️ *Atenção:* Custo de transporte elevado ({_fmt(price)}). Risco de desistência do hotel.",
            "📉 *Contorno:* Sugerir datas de partida +/- 2 dias. Frequentemente reduz custos em até 30%.",
            "🎁 *Retenção:* Oferecer crédito de consumo no resort (ex: R$ 300) para amortizar psicologicamente o custo do aéreo.",
        ]

    # Historical comparison
    hist = _analyse_historical(price, historical_prices or [])
    if hist["suggestion"]:
        suggestions.append(hist["suggestion"])

    if hist["percent_diff"] is not None and hist["percent_diff"] <= -15:
        if opportunity_level == "medium":
            opportunity_level = "high"
        elif opportunity_level == "low":
            opportunity_level = "medium"

    # Destination-specific tip
    tip = _destination_tip(destination)
    if tip:
        suggestions.append(tip)

    return {
        "opportunity_level": opportunity_level,
        "suggestions": suggestions,
        "price_analysis": {
            "price": price,
            "currency": "BRL",
            "thresholds": {"low": price_low, "medium": price_medium},
            "best_flight_price": (best_flight or {}).get("price"),
            "cheapest_flight_price": (cheapest_flight or {}).get("price"),
            "historical": hist,
        },
    }


# ── Helpers ──────────────────────────────────────────────────

def _analyse_historical(current_price: float, historical_prices: list) -> dict:
    valid = [float(h["price"]) for h in historical_prices if h.get("price") is not None]

    if not valid:
        return {"average_price": None, "percent_diff": None, "suggestion": None}

    avg = round(sum(valid) / len(valid))
    diff = round(((current_price - avg) / avg) * 100)

    if diff <= -15:
        suggestion = f"📉 Preço {abs(diff)}% abaixo da média histórica ({_fmt(avg)}) — ótima oportunidade!"
    elif diff <= -5:
        suggestion = f"📊 Preço {abs(diff)}% abaixo da média ({_fmt(avg)}) — momento favorável."
    elif diff >= 15:
        suggestion = f"📈 Preço {diff}% acima da média ({_fmt(avg)}) — sugerir datas alternativas."
    elif diff >= 5:
        suggestion = f"📊 Preço {diff}% acima da média ({_fmt(avg)}) — dentro da faixa normal."
    else:
        suggestion = None

    return {"average_price": avg, "percent_diff": diff, "suggestion": suggestion, "sample_size": len(valid)}


def _destination_tip(iata: str) -> str | None:
    tips = {
        "MIA": "🌴 Miami — sugerir passeio de barco em Biscayne Bay.",
        "MCO": "🎢 Orlando — oferecer ingressos de parques como add-on.",
        "CDG": "🗼 Paris — sugerir experiência gastronômica ou cruzeiro no Sena.",
        "FCO": "🏛️ Roma — oferecer tour guiado pelo Vaticano.",
        "LIS": "🇵🇹 Lisboa — sugerir passeio de tuk-tuk + degustação de vinhos.",
        "CUN": "🏖️ Cancún — sugerir pacote all-inclusive com upgrade.",
        "DXB": "🕌 Dubai — oferecer desert safari ou passeio de helicóptero.",
        "JFK": "🗽 Nova York — oferecer ingressos para Broadway.",
        "LHR": "🇬🇧 Londres — sugerir afternoon tea experience.",
    }
    return tips.get((iata or "").upper())


def _fmt(value) -> str:
    """Manual BRL formatter (no locale dependency)."""
    try:
        v = float(value)
        s = f"{v:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
        return f"R$ {s}"
    except (TypeError, ValueError):
        return "N/D"
